pkg/operator/rbaccontroller/rbaccontroller.go

```diff
@@ -0,0 +1,145 @@
+package rbaccontroller
+
+import (
+	"context"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	listerrbacv1 "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const (
+	// This annotation is set to a value of "true" on RBAC resources created by the openshift-apiserver.
+	openShiftManagedRBACAnnotation = "openshiftapiserver.rbac.authorization.openshift.io/managed"
+)
+
+var (
+	// The named cluster role bindings are created by openshift-apiserver
+	// and in previous releases may have included the unauthenticated subject.
+	clusterRoleBindingNames = []string{"discovery", "basic-users", "system:openshift:discovery",
+		"cluster-status-binding"}
+)
+
+// RBACController reconciles RBAC resources created by openshift-apiserver.
+//
+// If openshiftManagedRBACAnnotation is set to "true" for a cluster role binding,
+// RBACController ensures the removal of the unauthenticated subject from its list of subjects.
+//
+// A user can opt-out of this reconciliation by setting the annotation to "false".
+type RBACController struct {
+	eventRecorder            events.Recorder
+	kubeClient               kubernetes.Interface
+	clusterRoleBindingLister listerrbacv1.ClusterRoleBindingLister
+}
+
+func NewRBACController(kubeClient kubernetes.Interface, informerFactory informers.SharedInformerFactory, recorder events.Recorder) factory.Controller {
+	c := &RBACController{
+		kubeClient:               kubeClient,
+		clusterRoleBindingLister: informerFactory.Rbac().V1().ClusterRoleBindings().Lister(),
+		eventRecorder:            recorder,
+	}
+
+	syncCtx := factory.NewSyncContext("RBACController", c.eventRecorder)
+
+	informer := informerFactory.Rbac().V1().ClusterRoleBindings().Informer()
+	informer.AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: hasMatchingNames,
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err == nil {
+					syncCtx.Queue().Add(key)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(newObj)
+				if err == nil {
+					syncCtx.Queue().Add(key)
+				}
+			},
+		},
+	})
+
+	return factory.New().
+		WithBareInformers(informer).
+		WithSyncContext(syncCtx).
+		WithSync(c.sync).
+		ToController("", nil)
+}
+
+// sync ensures that cluster role bindings created by openshift-apiserver do not have the unauthenticated subject.
+func (c *RBACController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	var errs []error
+
+	for _, clusterRoleBindingName := range clusterRoleBindingNames {
```

Comment:
With the name and annotation filtering in place, it is no longer necessary to iterate over target names or check for the annotation. `syncContext` provides a key identifying the CRB that can be used to retrieve the resource from the informer cache.

Reply:
I tried using informers and filters but not sure if I am using them the proper way. So, It looks like correct :smile: 

I updated the code based on your comments. Please take a look!